emma-spark/src/main/scala/org/emmalanguage/api/SparkDataset.scala

```diff
@@ -116,6 +117,16 @@ class SparkDataset[A: Meta] private[api](@transient private[emmalanguage] val re
     rep.collect()
 
   // -----------------------------------------------------
+  // Pre-defined folds
+  // -----------------------------------------------------
+
+  override def find(p: A => Boolean): Option[A] =
+    Option(fold[Tuple1[A]](Tuple1(null.asInstanceOf[A]))(
+      x => if (p(x)) Tuple1(x) else Tuple1(null.asInstanceOf[A]),
+      (x, y) => if (x._1 != null) x else y
+    )._1)
```

Comment:
I implemented both methods directly on the underlying `Dataset` representation.

```scala
override def reduceOption(p: (A, A) => A): Option[A] =
  try {
    Option(rep.reduce(p))
  } catch {
    case e: UnsupportedOperationException if e.getMessage == "empty collection" => None
    case e: Throwable => throw e
  }

override def find(p: A => Boolean): Option[A] =
  try {
    Option(rep.filter(p).head())
  } catch {
    case e: NoSuchElementException if e.getMessage == "next on empty iterator" => None
    case e: Throwable => throw e
  }
```

The corner cases are covered with some additional tests in `DataBagSpec`.